### PR TITLE
NTP Sponsored Images: Fix crash when using component with no data

### PR DESCRIPTION
**Pull request by petemill**
- Adds validity checker to wallpaper-is-active checker
- Adds tests for this
- Validity checker checks wallpapers _and_ logo destination url
- Moves NTP sponsored images preferences to component so tests can use local component pref names

Fix https://github.com/brave/brave-browser/issues/7928

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:
On https://github.com/brave/brave-browser/issues/7928

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: components/ntp_sponsored_images/browser/view_counter_service.cc [hunk 1]
```diff
@@ -113,7 +116,8 @@ void ViewCounterService::RegisterPageView() {
 }
 
 bool ViewCounterService::IsBrandedWallpaperActive() {
-  return (is_supported_locale_ && IsOptedIn() && current_wallpaper());
+  return (is_supported_locale_ && IsOptedIn() && current_wallpaper() &&
+      current_wallpaper()->IsValid());
```
Comment (petemill) [inline components/ntp_sponsored_images/browser/view_counter_service.cc]: This line fixes the issue for this PR. The others all just support adding the test.

File: components/ntp_sponsored_images/browser/view_counter_service_unittest.cc [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+  void OptIn() {
+    prefs()->SetBoolean(
+        ntp_sponsored_images::prefs::kNewTabPageShowBrandedBackgroundImage,
+        true);
+  }
+
+  NTPSponsoredImagesData* CreateGoodData() {
+    NTPSponsoredImagesData* data = new NTPSponsoredImagesData;
+    // static auto demo = std::make_unique<NTPSponsoredImagesData>();
```
Comment (simonhong) [inline components/ntp_sponsored_images/browser/view_counter_service_unittest.cc]: I think this `make_unique` will work.

File: components/ntp_sponsored_images/browser/view_counter_service_unittest.cc [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    data->url_prefix = "chrome://newtab/ntp-dummy-brandedwallpaper/";
+    data->wallpaper_image_files = {
+        base::FilePath(FILE_PATH_LITERAL("wallpaper1.jpg")),
+        base::FilePath(FILE_PATH_LITERAL("wallpaper2.jpg")),
+        base::FilePath(FILE_PATH_LITERAL("wallpaper3.jpg")),
+    };
+    data->logo_alt_text = "Technikke: For music lovers.";
+    data->logo_company_name = "Technikke";
+    data->logo_destination_url = "https://brave.com";
+    // return demo.get();
```
Comment (simonhong) [inline components/ntp_sponsored_images/browser/view_counter_service_unittest.cc]: ditto.

File: components/ntp_sponsored_images/browser/view_counter_service.cc [hunk 4]
```diff
@@ -48,7 +49,8 @@ ViewCounterService::ViewCounterService(
       is_supported_locale_(is_supported_locale) {
   // If we have a wallpaper, store it as private var.
   // Set demo wallpaper if a flag is set.
-  if (!base::FeatureList::IsEnabled(features::kBraveNTPBrandedWallpaperDemo)) {
+  if (!base::FeatureList::IsEnabled(features::kBraveNTPBrandedWallpaperDemo) &&
+          service_) {
```
Comment (simonhong) [inline components/ntp_sponsored_images/browser/view_counter_service.cc]: Why did you check |service_| is null?
I can't find anyplace that passing null service.

File: browser/ntp_sponsored_images/view_counter_service_factory.cc [hunk 5]
```diff
@@ -68,11 +70,8 @@ KeyedService* ViewCounterServiceFactory::BuildServiceInstanceFor(
 }
 
 void ViewCounterServiceFactory::RegisterProfilePrefs(
-    user_prefs::PrefRegistrySyncable* registry) {
-  registry->RegisterBooleanPref(
-      kBrandedWallpaperNotificationDismissed, false);
-  registry->RegisterBooleanPref(
-      kNewTabPageShowBrandedBackgroundImage, true);
+     user_prefs::PrefRegistrySyncable* registry) {
+  RegisterSponsoredImagesProfilePrefs(registry);
```
Comment (bridiver) [inline browser/ntp_sponsored_images/view_counter_service_factory.cc]: can we use the common pattern of a static `RegisterProfilePrefs` method on the service?

**Comment by petemill** [comment]
CI passed with https://travis-ci.org/brave/brave-core/builds/642661555?utm_source=github_status&utm_medium=notification

File: components/ntp_sponsored_images/browser/view_counter_service_unittest.cc (hunk 2 above)
Comment (petemill) [reply components/ntp_sponsored_images/browser/view_counter_service_unittest.cc]: done

File: components/ntp_sponsored_images/browser/view_counter_service_unittest.cc (hunk 3 above)
Comment (petemill) [reply components/ntp_sponsored_images/browser/view_counter_service_unittest.cc]: done

File: components/ntp_sponsored_images/browser/view_counter_service.cc (hunk 4 above)
Comment (petemill) [reply components/ntp_sponsored_images/browser/view_counter_service.cc]: changed to DCHECK

File: browser/ntp_sponsored_images/view_counter_service_factory.cc (hunk 5 above)
Comment (petemill) [reply browser/ntp_sponsored_images/view_counter_service_factory.cc]: done

File: components/ntp_sponsored_images/browser/view_counter_service.cc [hunk 6]
```diff
@@ -71,9 +83,10 @@ ViewCounterService::~ViewCounterService() = default;
 NTPSponsoredImagesData* ViewCounterService::current_wallpaper() {
   if (base::FeatureList::IsEnabled(features::kBraveNTPBrandedWallpaperDemo)) {
     return GetDemoWallpaper();
-  } else {
+  } else if (service_) {
```
Comment (simonhong) [inline components/ntp_sponsored_images/browser/view_counter_service.cc]: I think this can be deleted?

Comment (petemill) [reply components/ntp_sponsored_images/browser/view_counter_service.cc]: done

**Comment by bsclifton** [comment]
CI failed on iOS for some reason, during init? Going to restart CI only for that platform. Otherwise, ready to merge 😄👍

**Pull request merged**